cloudinit/url_helper.py

```diff
@@ -59,7 +61,137 @@ def combine_single(url, add_on):
+    when not using strict mode (ftp://), fall back to using unencrypted ftp
+
+    returns: UrlResponse
+    """
+
+    def get_return_code_from_exception(exc):
+        # ftplib doesn't expose error codes, so use this lookup table
+        ftp_error_codes = {
+            ftplib.error_reply: 300,  # unexpected [123]xx reply
+            ftplib.error_temp: 400,  # 4xx errors
+            ftplib.error_perm: 500,  # 5xx errors
+            ftplib.error_proto: 600,  # response does not begin with [1-5]
+            EOFError: 700,  # made up
+            # OSError is also possible. Use OSError.errno for that.
+        }
+        code = ftp_error_codes.get(type(exc))  # pyright: ignore
+        if not code:
+            if isinstance(exc, OSError):
+                code = exc.errno
+            else:
+                LOG.warning(
+                    "Unexpected exception type while connecting to ftp server."
+                )
+        return code
+
+    url_parts = urlsplit(url)
+    if not url_parts.hostname:
+        raise UrlError(
+            cause="Invalid url provided", code=NOT_FOUND, headers=None, url=url
+        )
+    with io.BytesIO() as buffer:
+        port = url_parts.port or 21
+        user = url_parts.username or "anonymous"
+        try:
+            ftp_tls = ftplib.FTP_TLS(
+                context=create_default_context(),
+            )
+            LOG.debug(
+                "Attempting to connect to %s via port [%s] over tls.",
+                url, port
+            )
+            ftp_tls.connect(
+                host=url_parts.hostname,
+                port=port,
+                timeout=timeout or 0.0,  # Python docs are wrong about types
+            )
+            LOG.debug("Attempting to login with user [%s]", user)
+            ftp_tls.login(
+                user=user,
+                passwd=url_parts.password or "",
+            )
+            LOG.debug("Creating a secure connection", user)
+            ftp_tls.prot_p()
+            LOG.debug("Reading file: %s", url_parts.path)
+            ftp_tls.retrbinary(f"RETR {url_parts.path}", callback=buffer.write)
+            response = FtpResponse(url_parts.path, contents=buffer)
+            LOG.debug("Closing connection", url_parts.path)
+            ftp_tls.close()
+            return response
+        except ftplib.all_errors as e:
+            code = get_return_code_from_exception(e),
+            if "ftps" == url_parts.scheme:
+                raise UrlError(
+                    cause=(
+                        "Connecting to ftp server over tls "
+                        f"failed for url {url} [{code}]"
+                    ),
+                    code=code,
+                    headers=None,
+                    url=url,
+                ) from e
+            LOG.info(
+                "Connecting to ftp server over tls "
+                "failed for url %s [%s]", url, code
+            )
+        try:
```

Comment:
I'm wondering if there's a possible security concern here. If somebody has an ftps server, but then specifies "ftp://", will it still connect? If so, they may think they have a secure connection, but a MITM could host a different ftp server and trigger this fallback code.

Reply:
> I'm wondering if there's a possible security concern here. If somebody has an ftps server, but then specifies "ftp://", will it still connect?

By default, yes.

I believe this can be prevented by the user by configuring the base image with CA certs which don't trust the attacker's certs.    

> If so, they may think they have a secure connection, but a MITM could host a different ftp server and trigger this fallback code.

This is possible regardless of the fallback code (via "ftps://"). If an attacker can MITM the instance (and the instance's CA chain trusts the attacker's TLS certs), then the instance will have a TLS-encrypted connection to the MITM server which could itself host an ftp server.

I don't believe that the security concern mentioned here is any different than connecting via HTTPS, which we already support in this codepath. Both can be avoided by only initializing over trusted networks or (I think) using a custom CA.

Signed user-data is something we could implement someday to help these scenarios.

This is all probably better answered by @setharnold.

Reply:
> I believe this can be prevented by the user by configuring the base image with CA certs which don't trust the attacker's certs.
If an attacker can MITM the instance (and the instance's CA chain trusts the attacker's TLS certs), then the instance will have a TLS-encrypted connection to the MITM server which could itself host an ftp server.

But the attacker wouldn't need certs at all. If I put up a plain unencrypted, no-cert ftp server between cloud-init and the fully secure target ftps server, but specify `ftp://`, our code will see the MITM, attempt to connect securely, fail on the (normally used) ftps connection, log a message, then fallback to an unsecured ftp connect to the MITM. Right?

The obvious answer is "don't connect with ftp:// when you mean ftps://", but I'm just concerned that this is a potential security footgun.

Regardless, I agree it's best to check with security.

Reply:
> But the attacker wouldn't need certs at all.

Correct, that is how unencrypted ftp would work regardless of this fallback path.

The user explicitly asked to support an insecure unencrypted ftp by setting "ftp://", so attempting to use TLS first isn't introducing any vulnerabilities - it simply attempts a more secure connection first in case the server supports TLS. The "fallback" is to do exactly what the user asked.

> The obvious answer is "don't connect with ftp:// when you mean ftps://", but I'm just concerned that this is a potential security footgun.

The alternative is "cloud-init shouldn't support unencrypted ftp", and to quote your earlier comment on this PR:

> I'm not sure I agree about vanilla ftp. Isn't it still commonly used over internal networks? In browsers, support doesn't really buy much functionality other than not having to use another client, so it doesn't justify the maintenance burden. In cloud-init, it could potentially be a lot more useful, and the maintenance is few lines of fallback code.
>
>  it could potentially be a lot more useful, and the maintenance is few lines of fallback code.

Do you think we should support unencrypted ftp or don't you?

I'm in favor of it, since users may want it for use on a secure private management network. I think that providing the option, documenting the implications, and then allowing the user to define their own security posture makes sense. I also think that it doesn't do any harm to try ftps first, just in case it is supported by the server.

Reply:
> Do you think we should support unencrypted ftp or don't you?

You're framing this as if I just decided I want two totally opposite things. As I reviewed the code more carefully, some security implications stood out to me. I want to make sure this won't be a problem. Yes, I still want ftp...but I don't want it if it exposes a security vulnerability. I'm trying to figure that out and I'd rather not sweep those concerns under the rug. False dichotomies won't help the process.

> The alternative is "cloud-init shouldn't support unencrypted ftp"

Is that the only alternative? What if we required `ftps://` to connect to a secure server? That way there's no pretense that they'll be getting a secure connection when they won't be.

Reply:
> You're framing this as if I just decided I want two totally opposite things. As I reviewed the code more carefully, some security implications stood out to me. I want to make sure this won't be a problem.

My apologies, I'm not trying to shut down the conversation. It's good to talk through and make sure we understand the implementation details and users' understanding of the tool.

Ftp is not a secure transfer protocol and if we include it will have to be documented as such, regardless of whether we quietly attempt TLS for a more secure connection.

>> The alternative is "cloud-init shouldn't support unencrypted ftp"
>
> Is that the only alternative? What if we required ftps:// to connect to a secure server? That way there's no pretense that they'll be getting a secure connection when they won't be.

I think we were thinking about this in different ways. In your last comment I thought you were concerned that using TLS even when the user didn't request it would somehow make it easier for an attacker to intercept communication than normal unencrypted ftp would. Requiring `ftps://` to use TLS wouldn't make things any technically more secure, so I guess this is still about user perception of this protocol? I haven't put together docs yet, but if I think I would say something like:

| scheme | title | description |
| ------ | ----- | ----------- |
| ftps:// | FTP over TLS | acquire configurations from a server with TLS support |
| ftp:// | unencrypted FTP | not safe for use over untrusted networks, use at your own risk |

Reply:
> Requiring ftps:// to use TLS wouldn't make things any technically more secure, so I guess this is still about user perception of this protocol?

I think so? At the end of the day, I think you doc change would be good enough, but I think there's still a potential footgun there that's worth addressing.

I think it's a footgun because we're upgrading the connection to TLS, but in reality, it can only provide encryption, not identity. Somebody can look at the docs once, see we provide fpt(s) support, standup an ftps server, connect using "ftp://" and think they're secure because they see a TLS connection. Our fallback behavior prevents identity (because it can be MITMed), so "upgrading" to TLS without the possibility of real identity gives a false sense of security. At the end of the day we can say "you didn't RTFM", but if we could prevent that by only attempting a TLS connection when we see "ftps", then I think we should. I say that not really knowing the library though, so I don't actually know how hard that would be.

Reply:
fair enough, I've made the ftp:// scheme only support ftp, and the ftps:// scheme only support ftp over tls